docs(tip-1060): key credit accounting on present/new transitions (#5674)
tempoxyz/tempo: tips/tip-1060.md

1. **Storage deletion refunds are replaced by credits.** The legacy nonzero-to-zero clearing refund (`SSTORE_CLEARS_SCHEDULE` from EIP-3529) is set to zero. Instead, every nonzero-to-zero storage transition mints one storage credit for the account whose storage was modified. Other baseline SSTORE refunds, such as restore-to-original refunds from EIP-2200's net gas metering, are preserved.

2. **The TIP-1000 storage creation cost is split.** TIP-1000's storage creation component (`SSTORE_CREATE_COST`, 250,000 gas) becomes a residual (`SSTORE_SET_COST`, 5,000 gas) plus a creditable portion (`STORAGE_CREDIT_VALUE`, 245,000 gas). The residual is charged by the SSTORE gas function on clean creations only. The creditable portion is either paid as gas or covered by consuming one storage credit, depending on the account's selected mode.

3. **The EIP-3529 refund cap is removed.** The EIP-3529 limit that caps gas refunds to one fifth of gas used no longer applies. Storage-credit settlement refunds and preserved baseline SSTORE refunds are credited to the transaction's refund counter in full.

> A new Storage Credits precompile lets each account read its credit balance and choose how credits are used for its own storage creations. `Refund` mode is the default: creations pay `STORAGE_CREDIT_VALUE` upfront, and credits are matched at end-of-transaction settlement. This lets credits minted later in the same transaction refund earlier creations while keeping the transaction's required gas limit independent of the account's credit balance at inclusion time. `Preserve` mode leaves credits untouched, and `Direct` mode consumes available credits synchronously, optionally bounded by a transaction-local budget.

Reviewer: ```suggestion
A new Storage Credits precompile lets each account read its credit balance and choose how credits are used for its own storage creations. `Refund` mode is the default: creations pay `STORAGE_CREDIT_VALUE` upfront, and credits are matched at end-of-transaction settlement. This lets credits minted later in the same transaction refund earlier creations while keeping the transaction's required gas limit independent of the account's credit balance at inclusion time. `Preserve` mode does not consume credits, and `Direct` mode consumes available credits synchronously, optionally bounded by a transaction-local budget.
```

Traces and debug tooling SHOULD expose credit minting, synchronous `Direct` credit consumption, `Refund` pending counter increments, Direct budget reads and writes, end-of-transaction settlement (per-account credits consumed and refund credited), and storage creation mode reads or writes so developers can explain gas differences. Smart contracts that perform storage churn may observe lower effective costs after this TIP, including for transactions that net to zero new storage. SSTORE semantics for accounts that do not churn storage remain unchanged.

---

# Invariants

These invariants apply to storage operations governed by TIP-1060 and are evaluated over committed transaction effects after end-of-transaction settlement. Reverted execution scopes have no effect for invariant purposes.

For this section, a **backing unit** is an account-local entitlement corresponding to one 230,000-gas creditable storage-creation portion that has been paid and has not been returned through any refund/refill path. A backing unit may be represented by live nonzero storage owned by the account or by one unit of `storage_credit_balance`.

> 1. **Credit backing and conservation:** For each account `A`, every unit counted in `storage_credit_balance[A]` MUST be backed by one unique backing unit for `A`. Backing units may move between live nonzero storage owned by `A` and `storage_credit_balance[A]`, and may be retired when the corresponding 230,000-gas value is refunded or when a saturated credit balance cannot represent another released unit. Backing units MUST NOT be created without a paid, non-returned 230,000-gas creditable portion, duplicated, transferred between accounts, or used to back more than one live slot or represented storage credit.

Reviewer: won't fix

> 1. **Credit backing and conservation:** For each account `A`, every unit counted in `storage_credit_balance[A]` MUST be backed by one unique backing unit for `A`. Backing units may move between live nonzero storage owned by `A` and `storage_credit_balance[A]`, and may be retired when the corresponding 230,000-gas value is refunded or when a saturated credit balance cannot represent another released unit. Backing units MUST NOT be created without a paid, non-returned 230,000-gas creditable portion, duplicated, transferred between accounts, or used to back more than one live slot or represented storage credit.

Author: This is fine

2. **Zero-crossing storage transitions preserve backing:** Every committed storage transition owned by account `A` from zero to nonzero MUST leave the resulting live slot backed by exactly one account-local backing unit, either by paying the 230,000-gas creditable portion or by consuming an already-backed storage credit. Every committed transition from nonzero to zero releases the live slot's backing into at most one represented storage credit for `A`; if `storage_credit_balance[A]` is saturated, the backing may be discarded but MUST NOT be duplicated. Transitions that do not cross the zero/nonzero boundary do not create, consume, or release backing.

> 3. **Refund exclusivity:** A 230,000-gas creditable portion may be returned only by consuming one backed storage credit through TIP-1060 settlement. After the return, the refunded payment no longer counts as backing; the consumed credit's backing replaces it wherever needed to satisfy live-storage backing for the same account, or is retired if no such backing is needed. The same 230,000-gas payment MUST NOT both be returned as gas and leave behind a represented storage credit or additional live-slot backing. No other refund/refill mechanism may return the 230,000-gas creditable portion for storage-credit-governed writes.

Reviewer: amsterdam/ tip-1016 are inactive

> 3. **Refund exclusivity:** A 230,000-gas creditable portion may be returned only by consuming one backed storage credit through TIP-1060 settlement. After the return, the refunded payment no longer counts as backing; the consumed credit's backing replaces it wherever needed to satisfy live-storage backing for the same account, or is retired if no such backing is needed. The same 230,000-gas payment MUST NOT both be returned as gas and leave behind a represented storage credit or additional live-slot backing. No other refund/refill mechanism may return the 230,000-gas creditable portion for storage-credit-governed writes.

Author: This issue seems all over the place, t4 is always enabled and it does not effect tip1060 on sstore
